cpp/src/arrow/ipc/reader.cc

```diff
@@ -684,7 +685,19 @@ Status ReadDictionary(const Buffer& metadata, DictionaryMemo* dictionary_memo,
     return Status::Invalid("Dictionary record batch must only contain one field");
   }
   auto dictionary = batch->column(0);
-  return dictionary_memo->AddDictionary(id, dictionary);
+  if (dictionary_batch->isDelta()) {
+    std::shared_ptr<Array> originalDict, combinedDict;
+    RETURN_NOT_OK(dictionary_memo->GetDictionary(id, &originalDict));
+    ArrayVector dictsToCombine{originalDict, dictionary};
+    ARROW_ASSIGN_OR_RAISE(combinedDict, Concatenate(dictsToCombine, options.memory_pool));
+    return dictionary_memo->UpdateDictionary(id, combinedDict);
+  }
+
+  if (dictionary_memo->HasDictionary(id)) {
```

Comment:
Instead of spelling this out manually here, perhaps add `ReplaceDictionary` method to `DictionaryMemo`?

Reply:
Thanks for the good suggestion. I have added the AddOrReplaceDictionary method.